### PR TITLE
Improve S1244: Add message to use "IsNaN" instead of "== double.NaN"

### PR DESCRIPTION
**Pull request by antonioaversa**
Part of #6521

~~Made into a draft because it should be merged after the release.~~

### CODE REVIEW

File: analyzers/src/SonarAnalyzer.CSharp/Rules/EqualityOnFloatingPoint.cs [hunk 1]
```diff
@@ -85,12 +85,21 @@ private static void CheckEquality(SonarSyntaxNodeReportingContext context)
...
             }
         }
 
+        private static string ProposedMessage(SonarSyntaxNodeReportingContext context, ExpressionSyntax expression) =>
+            expression is MemberAccessExpressionSyntax memberAccess
+            && memberAccess.GetName() is var memberName
+            && FloatingPointEpsilonDependantMembers.TryGetValue(memberName, out var proposedMethod)
+            && context.SemanticModel.GetSymbolInfo(memberAccess.Expression).Symbol is ITypeSymbol type
+            && IsFloatingPointNumberType(type)
+            ? $"{type}.{proposedMethod}"
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/EqualityOnFloatingPoint.cs:100]: Use [`ToMinimalDisplayString`](https://learn.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.isymbol.tominimaldisplaystring?view=roslyn-dotnet-3.9.0)

File: analyzers/src/SonarAnalyzer.CSharp/Rules/EqualityOnFloatingPoint.cs [hunk 2]
```diff
@@ -85,12 +85,21 @@ private static void CheckEquality(SonarSyntaxNodeReportingContext context)
...
+                context.ReportIssue(Diagnostic.Create(Rule, equals.OperatorToken.GetLocation(), messageEqualityPart, proposed));
             }
         }
 
+        private static string ProposedMessage(SonarSyntaxNodeReportingContext context, ExpressionSyntax expression) =>
+            expression is MemberAccessExpressionSyntax memberAccess
+            && memberAccess.GetName() is var memberName
+            && FloatingPointEpsilonDependantMembers.TryGetValue(memberName, out var proposedMethod)
+            && context.SemanticModel.GetSymbolInfo(memberAccess.Expression).Symbol is ITypeSymbol type
+            && IsFloatingPointNumberType(type)
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/EqualityOnFloatingPoint.cs:99]: Cleanup, simplified name with same meaning
```suggestion
            && IsFloatingPointType(type)
```

File: analyzers/src/SonarAnalyzer.CSharp/Rules/EqualityOnFloatingPoint.cs [hunk 3]
```diff
@@ -101,33 +110,29 @@ private static void CheckEquality(SonarSyntaxNodeReportingContext context)
...
+            && HasFloatingType(semanticModel, right.Left, right.Right);
 
         private static bool IsIndirectEquality(SemanticModel semanticModel, BinaryExpressionSyntax binaryExpression, BinaryExpressionSyntax left, BinaryExpressionSyntax right) =>
             binaryExpression.IsKind(SyntaxKind.LogicalAndExpression)
-                && HasAppropriateOperatorsForEquality(left, right)
-                && HasFloatingType(semanticModel, right.Left, right.Right);
+            && IsOperatorPair(left, right, SyntaxKind.GreaterThanEqualsToken, SyntaxKind.LessThanEqualsToken)
+            && HasFloatingType(semanticModel, right.Left, right.Right);
 
         private static bool HasFloatingType(SemanticModel semanticModel, ExpressionSyntax left, ExpressionSyntax right) =>
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/EqualityOnFloatingPoint.cs:128]: Both invocations are decomposing binary into left/right anyway
```suggestion
        private static bool HasFloatingType(SemanticModel semanticModel, BinaryExpressionSyntax binary) =>
```

**Review by pavel-mikula-sonarsource — CHANGES_REQUESTED** [review]
Minor improvements

File: analyzers/src/SonarAnalyzer.CSharp/Rules/EqualityOnFloatingPoint.cs [hunk 4]
```diff
@@ -24,15 +24,19 @@ namespace SonarAnalyzer.Rules.CSharp
...
+        private static readonly DiagnosticDescriptor Rule = DescriptorFactory.Create(DiagnosticId, MessageFormat);
 
-        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(rule);
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(Rule);
 
         private static readonly ISet<string> EqualityOperators = new HashSet<string> { "op_Equality", "op_Inequality" };
 
+        private static readonly IDictionary<string, string> FloatingPointEpsilonDependantMembers = new Dictionary<string, string>
+        {
+            [nameof(double.NaN)] = $"{nameof(double.IsNaN)}()",
```
[the remark this answers is not in the document]
Comment (antonioaversa) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/EqualityOnFloatingPoint.cs:37]: Interested in understanding why.

Initializer syntax with braces is the one at the collection level, whereas the dictionary-specific syntax sugar (using `[]`) was introduced later (C#6), since collection-based syntax was considered "somewhat suboptimal" (by Microsoft...).

c.p. https://learn.microsoft.com/en-us/archive/msdn-magazine/2014/special-issue/csharp-how-csharp-6-0-simplifies-clarifies-and-condenses-your-code#dictionary-initializer

File: analyzers/src/SonarAnalyzer.CSharp/Rules/EqualityOnFloatingPoint.cs (hunk 1 above)
Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/EqualityOnFloatingPoint.cs:100]: I would scope the alias definition to a single test case. That way the problems go away.
There might also be an option in the formatting setting on how to handle aliases so you could in theory turn off the aliasing but I wouldn't do that. That alias stuff is really unlikely to be an issue because no sane person would alias System.Double which already has an alias (the keyword double). The alias stuff makes more sense for [`Vector<T>`](https://learn.microsoft.com/en-us/dotnet/api/system.numerics.vector-1?view=net-7.0) and the like.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/EqualityOnFloatingPoint.CSharp11.cs [hunk 5]
```diff
@@ -4,6 +4,24 @@
 
 public class EqualityOnFloatingPoint
 {
+    void TestHalfNaN()
+    {
+        bool b;
+        var h = Half.NaN;
+
+        b = h == Half.NaN; // Noncompliant {{Do not check floating point equality with exact values, use 'Half.IsNaN()' instead.}}
```
Comment (martin-strecker-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/EqualityOnFloatingPoint.CSharp11.cs:12]: Add a test case with
```cs
using static System.Double;
_ = d == NaN;
```

You can add this test case in another file as well. I think it was introduced in C#6.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/EqualityOnFloatingPoint.cs [hunk 6]
```diff
@@ -2,86 +2,132 @@
...
 
-namespace Tests.Diagnostics
+public class EqualityOnFloatingPoint
 {
-    public class EqualityOnFloatingPoint
+    void Test(float f, double d1, double d2)
     {
-        void test(float f, double d1, double d2)
+        dynamic din = null;
+        if (din == null) { }
```
Comment (martin-strecker-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/EqualityOnFloatingPoint.cs:10]: Do we really need this? It doesn't seem to test anything relevant. If we want to keep it:

* Remove the local declaration and make it a parameter instead
* Rename the variable to `dyn`.
* Change the test case so that a floating point number is involved on one side of the binary.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/EqualityOnFloatingPoint.cs [hunk 7]
```diff
@@ -2,86 +2,132 @@
...
+        b = double.NaN != d;          // Noncompliant {{Do not check floating point inequality with exact values, use 'double.IsNaN()' instead.}}
+        b = double.NaN != double.NaN; // Noncompliant {{Do not check floating point inequality with exact values, use 'double.IsNaN()' instead.}}
+        b = iPar != double.NaN;       // Noncompliant {{Do not check floating point inequality with exact values, use 'double.IsNaN()' instead.}}
+        b = dPar != double.NaN;       // Noncompliant {{Do not check floating point inequality with exact values, use 'double.IsNaN()' instead.}}
+        b = fPar != double.NaN;       // Noncompliant {{Do not check floating point inequality with exact values, use 'double.IsNaN()' instead.}}
+        b = float.NaN != double.NaN;  // Noncompliant {{Do not check floating point inequality with exact values, use 'double.IsNaN()' instead.}}
+
+        // Wrong message: should be "use double.IsNaN()" instead
+        b = double.NaN != float.NaN;  // Noncompliant {{Do not check floating point inequality with exact values, use 'float.IsNaN()' instead.}}
```
Comment (martin-strecker-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/EqualityOnFloatingPoint.cs:77]: `SemanticModel.GetTypeInfo(node).ConvertedType` might help here https://learn.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.typeinfo

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/EqualityOnFloatingPoint.cs [hunk 8]
```diff
@@ -2,86 +2,132 @@
...
-            }
-
-            float f1 = 0.0F;
-            if ((System.Math.Sign(f1) == 0))
-            {
-            }
+            bool b;
+            DoubleAlias d1 = 1.674927471E-27;
+
+            b = d1 == DoubleAlias.NaN;    // Noncompliant {{Do not check floating point equality with exact values, use 'DoubleAlias.IsNaN()' instead.}}
```
Comment (martin-strecker-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/EqualityOnFloatingPoint.cs:127]:
```suggestion
            DoubleAlias d1 = 1.674927471E-27;

            var b = d1 == DoubleAlias.NaN;    // Noncompliant {{Do not check floating point equality with exact values, use 'DoubleAlias.IsNaN()' instead.}}
```

**Review by martin-strecker-sonarsource — CHANGES_REQUESTED** [review]
Looks good to me. There are some edge cases you may want to resolve but nothing you should spend to much time resolving because the issues are not blocking.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/EqualityOnFloatingPoint.CSharp11.cs (hunk 5 above)
Comment (martin-strecker-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/EqualityOnFloatingPoint.CSharp11.cs:12]: Add a test case where `T.IsNan()` is proposed:
```cs
public class C
{
    public void M<T>(T t) where T : IFloatingPointIeee754<T>
    {
        if (t == T.NaN) { } // Uncompliant
        if (T.IsNaN(t)) { } // Compliant
    }
}
```

Comment (antonioaversa) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/EqualityOnFloatingPoint.CSharp11.cs:12]: `using static System.Double` is a good one: thanks for spotting it out!

Although probably uncommon, I am going to try to have it work.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/EqualityOnFloatingPoint.cs [hunk 9]
```diff
@@ -24,15 +24,19 @@ namespace SonarAnalyzer.Rules.CSharp
     public sealed class EqualityOnFloatingPoint : SonarDiagnosticAnalyzer
```
[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/EqualityOnFloatingPoint.cs:24]: There's 👍 and the change wasn't done, so I'm not sure what the state is. Will you do it?

File: analyzers/src/SonarAnalyzer.CSharp/Rules/EqualityOnFloatingPoint.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/EqualityOnFloatingPoint.cs]: It's ugly. There are way too many different characters needed to make it work, while it does almost the same thing.

Functionally, I don't like the `[]` as it's a collection initializer that would not throw an exception in case of a duplicated key. And for collection initialization, it doesn't make sense to use duplicated keys.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/EqualityOnFloatingPoint.CSharp11.cs [hunk 10]
```diff
@@ -4,6 +4,24 @@
 
 public class EqualityOnFloatingPoint
 {
+    void testHalf()
```
[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/EqualityOnFloatingPoint.CSharp11.cs]: The point was also that the `Test` doesn't need to be there. Everything inside `TestCases` directory is a test.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/EqualityOnFloatingPoint.cs [hunk 11]
```diff
@@ -24,15 +24,19 @@ namespace SonarAnalyzer.Rules.CSharp
...
-        private static readonly DiagnosticDescriptor rule =
-            DescriptorFactory.Create(DiagnosticId, MessageFormat);
+        private static readonly DiagnosticDescriptor Rule = DescriptorFactory.Create(DiagnosticId, MessageFormat);
 
-        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(rule);
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(Rule);
 
         private static readonly ISet<string> EqualityOperators = new HashSet<string> { "op_Equality", "op_Inequality" };
 
+        private static readonly IDictionary<string, string> FloatingPointEpsilonDependantMembers = new Dictionary<string, string>
```
[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/EqualityOnFloatingPoint.cs]: I'd prefer `SpecialMembers`, as WellKnown like `double.MinValue` is fine and doesn't mean anything for this rule

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/EqualityOnFloatingPoint.CSharp11.cs [hunk 12]
```diff
@@ -4,6 +4,24 @@
 
 public class EqualityOnFloatingPoint
 {
+    void TestHalfNaN()
+    {
+        bool b;
+        var h = Half.NaN;
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/EqualityOnFloatingPoint.CSharp11.cs:10]: In this case, it's not clear to me that it raises because one of them is NaN, or both, or one is tracked as NaN and compared to the other. 

Let's remove `h` and add `Half arg` method parameter to express that any unknown value compared to the constant will trigger the rule.

Same everywhere

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/EqualityOnFloatingPoint.cs [hunk 13]
```diff
@@ -2,86 +2,132 @@
...
 using System.Runtime.Versioning;
 
-namespace Tests.Diagnostics
+public class EqualityOnFloatingPoint
 {
-    public class EqualityOnFloatingPoint
+    void Test(float f, double d1, double d2)
     {
-        void test(float f, double d1, double d2)
+        dynamic din = null;
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/EqualityOnFloatingPoint.cs:9]: Old code: It's weird that DYNamic is called DIN

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/EqualityOnFloatingPoint.cs [hunk 14]
```diff
@@ -2,86 +2,132 @@
...
+        if (d1 == d2) { }           // Noncompliant
+        if (d1 < 0 || d1 > 1) { }   // Compliant no indirect inequality test
+
+        var b = d1 == 3.14;                   // Noncompliant
+        if (true && f >= 3.146) { }           // Compliant no indirect equality test
+        if (f <= 3.146 && ((f >= 3.146))) { } // Noncompliant indirect equality test
+        if (3.146 >= f && 3.146 <= f) { }     // Noncompliant indirect equality test
+        if (f <= 3.146 && 3.146 <= f) { }     // FN. Equivalent to the case above but not detected
+        if (3.146 >= f && 3.146 < f) { }      // Compliant no indirect equality test
+        if (f <= 3.146 && f > 3.146) { }      // Compliant no indirect equality test
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/EqualityOnFloatingPoint.cs:23]: This is so much easier to follow and see what the cases really cover

File: analyzers/src/SonarAnalyzer.CSharp/Rules/EqualityOnFloatingPoint.cs [hunk 15]
```diff
@@ -23,15 +23,17 @@ namespace SonarAnalyzer.Rules.CSharp
...
-        private const string MessageFormat = "Do not check floating point {0} with exact values, use a range instead.";
+        private const string DiagnosticId = "S1244";
+        private const string MessageFormat = "Do not check floating point {0} with exact values, use {1} instead.";
 
-        private static readonly DiagnosticDescriptor rule =
-            DescriptorFactory.Create(DiagnosticId, MessageFormat);
-
-        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(rule);
+        private static readonly DiagnosticDescriptor Rule = DescriptorFactory.Create(DiagnosticId, MessageFormat);
+        private static readonly string[] EqualityOperators = new[] { "op_Equality", "op_Inequality" };
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/EqualityOnFloatingPoint.cs:30]: Small simplification in case Martin's suggestion doesn't work. The implementation checks for `MethodKind.UserDefinedOperator` that I'm not sure how it behaves
```suggestion
        private static readonly string[] EqualityOperators = { "op_Equality", "op_Inequality" };
```

File: analyzers/src/SonarAnalyzer.CSharp/Rules/EqualityOnFloatingPoint.cs [hunk 16]
```diff
@@ -49,85 +51,75 @@ protected override void Initialize(SonarAnalysisContext context)
...
-            {
-                return;
-            }
-
-            var isEquality = IsIndirectEquality(context.SemanticModel, binaryExpression, left, right);
 
-            if (isEquality || IsIndirectInequality(context.SemanticModel, binaryExpression, left, right))
+            if (TryGetBinaryExpression(binaryExpression.Left) is { } left
+                && TryGetBinaryExpression(binaryExpression.Right) is { } right
+                && (CSharpEquivalenceChecker.AreEquivalent(right.Right, left.Right) && CSharpEquivalenceChecker.AreEquivalent(right.Left, left.Left))
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/EqualityOnFloatingPoint.cs:57]: Having `&&` needs this formatting
```suggestion
                && CSharpEquivalenceChecker.AreEquivalent(right.Right, left.Right) 
                && CSharpEquivalenceChecker.AreEquivalent(right.Left, left.Left)
```

File: analyzers/src/SonarAnalyzer.CSharp/Rules/EqualityOnFloatingPoint.cs [hunk 17]
```diff
@@ -49,85 +51,75 @@ protected override void Initialize(SonarAnalysisContext context)
...
+                && IsIndirectEquality(context.SemanticModel, binaryExpression, left, right) is var isEquality
+                && IsIndirectInequality(context.SemanticModel, binaryExpression, left, right) is var isInequality
+                && (isEquality || isInequality))
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/EqualityOnFloatingPoint.cs:60]: I would normally ask to remove the isInequality because it's used only once, but in this case, it makes the code more symetrical (easy to follow and see the differences), so we'll keep it

**Review by pavel-mikula-sonarsource — CHANGES_REQUESTED** [review]
Almost there

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/EqualityOnFloatingPoint.cs (hunk 6 above)
Comment (antonioaversa) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/EqualityOnFloatingPoint.cs:10]: Indeed, we don't need this check in the way it currently is in the code.

However, applying your suggestions:

```cs
void Test(float f, double d1, double d2, dynamic dyn)
    {
        if (dyn == 3.14) { }        // Noncompliant {{Do not check floating point equality with exact values, use a range instead.}}
```

It becomes a false negative, as in the following example (which prints `False`):
```cs
dynamic x = 0.1 + 0.2;
var b = x == 0.3;
Console.WriteLine(b);
```

The problem comes from:
```cs
context.SemanticModel.GetSymbolInfo(equals).Symbol is IMethodSymbol { ContainingType: { } container, Name: { } equalitySymbolName }
```
where `Name` is correctly reported as `op_Equality`, but `ContainingType` is not (`null`).

If you agree, I would leave this as a FP, and not try to infer the true type of equality in presence of `dynamic`.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/EqualityOnFloatingPoint.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (antonioaversa) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/EqualityOnFloatingPoint.cs:24]: Did on the wrong file. Fixed now.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/EqualityOnFloatingPoint.CSharp11.cs (hunk 5 above)
Comment (antonioaversa) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/EqualityOnFloatingPoint.CSharp11.cs:12]: Easier than expected, it's fixed now.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/EqualityOnFloatingPoint.cs (hunk 15 above)
Comment (antonioaversa) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/EqualityOnFloatingPoint.cs:30]: Not needed, as `EqualityOperators` has been removed.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/EqualityOnFloatingPoint.cs (hunk 13 above)
Comment (antonioaversa) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/EqualityOnFloatingPoint.cs:9]: Fixed... and made it somewhat useful too.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/EqualityOnFloatingPoint.cs (hunk 8 above)
Comment (antonioaversa) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/EqualityOnFloatingPoint.cs:127]: Actually that `bool` variable was not needed, and added nothing to the test, so I have removed it

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/EqualityOnFloatingPoint.CSharp11.cs (hunk 12 above)
Comment (antonioaversa) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/EqualityOnFloatingPoint.CSharp11.cs:10]: It was meant to also check for compliance of `h = Half.NaN`, which is not a comparison but an assignment.

Made into a method parameter and added the `// Compliant`.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/EqualityOnFloatingPoint.cs (hunk 7 above)
Comment (antonioaversa) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/EqualityOnFloatingPoint.cs:77]: Great, fixed!

File: analyzers/src/SonarAnalyzer.CSharp/Rules/EqualityOnFloatingPoint.cs (hunk 15 above)
[the remark this answers is not in the document]
Comment (antonioaversa) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/EqualityOnFloatingPoint.cs]: Done.
It required a bit of refactoring with @pavel-mikula-sonarsource, as mentioned in https://github.com/SonarSource/sonar-dotnet/pull/6664#discussion_r1086563941

**Review by pavel-mikula-sonarsource — APPROVED** [review]
LGTM 🎉 I'll let @martin-strecker-sonarsource resolve his comments

**Review by martin-strecker-sonarsource — CHANGES_REQUESTED** [review]
Minor changes only

File: analyzers/src/SonarAnalyzer.CSharp/Rules/OperatorsShouldBeOverloadedConsistently.cs [hunk 18]
```diff
@@ -102,56 +102,35 @@ private static IEnumerable<string> FindMissingMethods(INamedTypeSymbol classSymb
...
-            }
-        }
+                _ when member.MethodKind != MethodKind.UserDefinedOperator => null,
+                _ when KnownMethods.IsOperatorBinaryPlus(member) => MethodName.OperatorPlus,
+                _ when KnownMethods.IsOperatorBinaryMinus(member) => MethodName.OperatorMinus,
+                _ when KnownMethods.IsOperatorBinaryMultiply(member) => MethodName.OperatorMultiply,
+                _ when KnownMethods.IsOperatorBinaryDivide(member) => MethodName.OperatorDivide,
+                _ when KnownMethods.IsOperatorBinaryModulus(member) => MethodName.OperatorReminder,
+                _ when KnownMethods.IsOperatorEquals(member) => MethodName.OperatorEquals,
+                _ when KnownMethods.IsOperatorNotEquals(member) => MethodName.OperatorNotEquals,
```
[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/OperatorsShouldBeOverloadedConsistently.cs:132]: Changing existing behavior of OperatorsShouldBeOverloadedConsistently.cs is out of scope for this PR

File: analyzers/src/SonarAnalyzer.CSharp/Rules/EqualityOnFloatingPoint.cs [hunk 19]
```diff
@@ -18,116 +18,119 @@
...
+            ? $"'{type.ToMinimalDisplayString(context.SemanticModel, memberAccess.SpanStart)}.{proposedMethod}()'"
+            : null;
+
+    private static string ProposedMessageForIdentifier(SonarSyntaxNodeReportingContext context, ExpressionSyntax expression) =>
+        expression is IdentifierNameSyntax identifier
+        && SpecialMembers.TryGetValue(identifier.GetName(), out var proposedMethod)
+        && context.SemanticModel.GetSymbolInfo(identifier).Symbol is { ContainingType: { } type }
+        && IsFloatingPointType(type)
+            ? $"'{proposedMethod}()'"
+            : null;
```
[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/EqualityOnFloatingPoint.cs:100]: I wouldn't do that. I took a look at it and it would make it less understandable

[the remark this answers is not in the document]
Comment (antonioaversa) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/EqualityOnFloatingPoint.cs:100]: Looks complex to make it simpler for me too: too many small differences to factorize it effectively:
- one uses `GetTypeInfo`, the other `GetSymbolInfo`
- one uses `ConvertedType`, the other `Symbol`, then gets its `ContainingType`, checking its non-nullability
- one makes use of the type in the generated string, calling `ToMinimalDisplayString`, the other doesn't

@martin-strecker-sonarsource if you have any suggestions (maybe using differently roslyin API), I'm keen to make the change.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/OperatorsShouldBeOverloadedConsistently.cs [hunk 20]
```diff
@@ -102,56 +102,35 @@ private static IEnumerable<string> FindMissingMethods(INamedTypeSymbol classSymb
...
+        }
 
-            if (classMethods.Any(KnownMethods.IsObjectGetHashCode))
+        private static string ImplementedOperator(IMethodSymbol member) =>
+            member switch
             {
-                yield return MethodName.ObjectGetHashCode;
-            }
-        }
+                _ when member.MethodKind != MethodKind.UserDefinedOperator => null,
```
[the remark this answers is not in the document]
Comment (antonioaversa) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/OperatorsShouldBeOverloadedConsistently.cs:125]: The rule [S4050](https://sonarsource.github.io/rspec/#/rspec/S4050/csharp) only focuses on methods that are user-defined.

This is the reason why `KnownMethods.IsOperatorXXX` originally had the constraint `MethodKind: MethodKind.UserDefinedOperator`.

By partially relaxing that constraint (which is now `MethodKind: MethodKind.BuiltinOperator or MethodKind.UserDefinedOperator)`, we could reuse `KnownMethods.IsOperatorXXX` elsewhere.

That, however, required adding an additional check here, to ensure operators defined at higher level in the class hierarchy would not be taken into account.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/OperatorsShouldBeOverloadedConsistently.cs (hunk 18 above)
[the remark this answers is not in the document]
Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/OperatorsShouldBeOverloadedConsistently.cs:132]: I didn't realize that this is something totally unrelated. I thought it is a helper needed by the original rule. I'll resolve as "Won't fix".

> If you agree with that, I would try to make the necessary changes.

Technically correct but too much of a niche for me to fix. See my comment above: I was in the wrong context while reviewing.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/OperatorsShouldBeOverloadedConsistently.cs (hunk 20 above)
[the remark this answers is not in the document]
Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/OperatorsShouldBeOverloadedConsistently.cs:125]: Okay. Got it now. You may still use the pattern syntax instead of the `when` if you like.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/EqualityOnFloatingPoint.cs (hunk 19 above)
[the remark this answers is not in the document]
Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/EqualityOnFloatingPoint.cs:100]: `GetTypeInfo` .. `ConvertedType` should also work for the identifier. So it might be possible to do it like so
`ProposedMessage(SemanticModel model, SyntaxNode(or ExpressionSyntaxNode) node, bool prefixType)` but I agree with Pavel that it might do more harm than good.

**Review by martin-strecker-sonarsource — APPROVED** [review]
LGTM. I left two suggestions open but you may decide on your own whether you want to fix it. They are just small refactoring suggestions.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/OperatorsShouldBeOverloadedConsistently.cs (hunk 20 above)
[the remark this answers is not in the document]
Comment (antonioaversa) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/OperatorsShouldBeOverloadedConsistently.cs:125]: As I don't have an opinion on whether the `!=` is better or worse than `: not` in this specific context, I've welcomed your suggestion and made the change.

I am curious though, about why would recommend the pattern-matching version over the classic way.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/EqualityOnFloatingPoint.cs (hunk 19 above)
[the remark this answers is not in the document]
Comment (antonioaversa) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/EqualityOnFloatingPoint.cs:100]: @pavel-mikula-sonarsource , @martin-strecker-sonarsource : Since you both seem to agree, can this be resolved? 

Or an action on my part is expected?

File: analyzers/src/SonarAnalyzer.CSharp/Rules/OperatorsShouldBeOverloadedConsistently.cs (hunk 20 above)
[the remark this answers is not in the document]
Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/OperatorsShouldBeOverloadedConsistently.cs]: Fairly simple: `when` is the fallback whenever pattern matching isn't possible.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/EqualityOnFloatingPoint.cs (hunk 19 above)
[the remark this answers is not in the document]
Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/EqualityOnFloatingPoint.cs:100]: No action needed. The suggestion was optional in case you wanted to give it a try. That's why I left it open for better visibility.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/OperatorsShouldBeOverloadedConsistently.cs (hunk 20 above)
[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/OperatorsShouldBeOverloadedConsistently.cs]: I agree that a pattern matching is better than the ugly workaround we did
